app/javascript/crayons/MarkdownToolbar/MarkdownToolbar.jsx

```diff
@@ -181,12 +182,36 @@ export const MarkdownToolbar = ({ textAreaId }) => {
   const insertSyntax = (syntaxName) => {
     setOverflowMenuOpen(false);
 
-    const { newTextAreaValue, newCursorStart, newCursorEnd } =
-      markdownSyntaxFormatters[syntaxName].getFormatting(textArea);
+    const {
+      newCursorStart,
+      newCursorEnd,
+      editSelectionStart,
+      editSelectionEnd,
+      replaceSelectionWith,
+    } = markdownSyntaxFormatters[syntaxName].getFormatting(textArea);
+
+    // We try to update the textArea with document.execCommand, which requires the contentEditable attribute to be true.
+    // The value is later toggled back to 'false'
+    textArea.contentEditable = 'true';
+    textArea.focus({ preventScroll: true });
+    textArea.setSelectionRange(editSelectionStart, editSelectionEnd);
+
+    try {
+      // We first try to use execCommand which allows the change to be correctly added to the undo queue.
+      // document.execCommand is deprecated, but the API which will eventually replace it is still incoming (https://w3c.github.io/input-events/)
+      document.execCommand('insertText', false, replaceSelectionWith);
```

Comment:
I've added a comment in the code too, but just to highlight that this is a deprecated function. The API due to replace still isn't in place though, and it seems like conventional wisdom at the moment to keep using `execCommand` for the purpose of maintaining the undo queue.

The [compatibility for this particular command](https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand#browser_compatibility) is very good, and to be honest, I've mostly been guided by github's implementation which utilises a try/catch like this for the same:

https://github.com/github/markdown-toolbar-element/blob/main/src/index.ts#L384-L392